File: ui/src/lib/services/worktrees.ts
import { api } from './api';

export async function createWorktree(projectId: string, data: { branch: string }): Promise<void> {
	await api.post(`projects/${projectId}/worktrees/create`, { json: data });
}

/** Rename the branch checked out in a worktree (`git branch -m`). Path and worktree id are unchanged. */
export async function renameWorktreeBranch(
	projectId: string,
	worktreeId: string,
	newBranch: string
): Promise<void> {
	const encoded = encodeURIComponent(worktreeId);
	await api.patch(`projects/${projectId}/worktrees/${encoded}`, { json: { newBranch } });
}

export async function deleteWorktree(projectId: string, worktreeId: string): Promise<void> {
	const encoded = encodeURIComponent(worktreeId);
	await api.delete(`projects/${projectId}/worktrees/${encoded}`);
}

export async function updateWorktreeExpandedState(
	projectId: string,
	worktreeId: string,
	expandedState: boolean
): Promise<void> {
	const encoded = encodeURIComponent(worktreeId);
	await api.patch(`projects/${projectId}/worktrees/${encoded}/expanded-state`, {
		json: { expandedState }
	});
}

File: src/commands/worktrees.rs
use tauri::State;

use crate::db::DbHandle;
use crate::error::AppError;
use crate::features::types::{CreateWorktreeRequest, RenameWorktreeBranchRequest};
use crate::features::worktrees::Service;

#[tauri::command]
#[specta::specta]
pub async fn create_worktree(
    db: State<'_, DbHandle>,
    project_id: String,
    input: CreateWorktreeRequest,
) -> Result<(), AppError> {
    Service::new(db.inner().clone())
        .create(&project_id, &input.branch)
        .await?;
    Ok(())
}

#[tauri::command]
#[specta::specta]
pub async fn delete_worktree(
    db: State<'_, DbHandle>,
    project_id: String,
    worktree_id: String,
) -> Result<(), AppError> {
    Service::new(db.inner().clone())
        .delete_checkout(&project_id, &worktree_id)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn rename_worktree_branch(
    db: State<'_, DbHandle>,
    project_id: String,
    worktree_id: String,
    input: RenameWorktreeBranchRequest,
) -> Result<(), AppError> {
    Service::new(db.inner().clone())
        .rename_checked_out_branch(&project_id, &worktree_id, &input.new_branch)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn update_worktree_expanded_state(
    db: State<'_, DbHandle>,
    project_id: String,
    worktree_id: String,
    expanded_state: bool,
) -> Result<(), AppError> {
    Service::new(db.inner().clone())
        .set_expanded_state(&project_id, &worktree_id, expanded_state)
        .await
}
